compiler/src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala

```diff
@@ -147,8 +147,11 @@ class IsInstanceOfEvaluator extends MiniPhaseTransform { thisTransformer =>
 
           val inMatch = s.qualifier.symbol is Case
 
-          if (valueClassesOrAny) tree
-          else if (knownStatically)
+          if (valueClassesOrAny) {
+            if (selector eq defn.ObjectType)
```

Comment:
Check the definition of `valueClassesOrAny`. If the selector is also `Object` here, we are dealing with a generic type param that's been erased.

We should also check to see that it was not originally Object either.